File: src/vertex.ts
import { Vector3D } from "./vector"
import { AttributeType } from "./attribute-info"

export default class Vertex{
    position: Vector3D
    color: number[] = [0,0,0, 1]
    constructor(position: Vector3D, color: number[] = [0,0,0, 1]){
       this.position = position
       this.color = color
    }
    attributeValues(): Map<string, number[]>{
        let attributes = new Map<string, number[]>()
        attributes.set(Vertex.positionAttribute, [this.position.x, this.position.y, this.position.z])
        attributes.set(Vertex.colorAttribute, this.color)
        return attributes
    }
    static size = 3
    static positionAttribute = "vertex_position"
    static colorAttribute = "vertex_color"
    static attributeMappings(): Map<string, {type: number, mapper: (gl, position) => void}>{
        let attributes = new Map<string, {type: number, mapper: (gl, position) => void}>()
        attributes.set(Vertex.positionAttribute, { type: AttributeType.ARRAY_BUFFER, mapper:  (gl,position) => gl.vertexAttribPointer(position, Vertex.size, gl.FLOAT, false, 0, 0) });
        attributes.set(Vertex.colorAttribute, { type: AttributeType.ARRAY_BUFFER, mapper:  (gl,position) => gl.vertexAttribPointer(position, 4, gl.FLOAT, false, 0, 0) });
        return attributes
    }
    
}

File: src/main/kotlin/simpleGl/Vertex.kt
package simpleGl

import org.khronos.webgl.WebGLRenderingContext


data class Vertex3D(
        val position: Vector3 = Vector3(x = 0f, y = 0f, z = 0f),
        val color: Color = Color(r = 0f, g = 0f, b = 0f, a = 1f)
){
    fun attributes() = mapOf( "vertex_position" to position.components, "vertex_color" to color.components )

    companion object {
        fun attributeMappings() = listOf(
                Attribute("vertex_position", AttributeType.ARRAY_BUFFER) { context, position ->
                    context.vertexAttribPointer(position, 3, WebGLRenderingContext.FLOAT, false, 0, 0)
                },
                Attribute("vertex_color", AttributeType.ARRAY_BUFFER) { context, position ->
                    context.vertexAttribPointer(position, 4, WebGLRenderingContext.FLOAT, false, 0, 0)
                }
        )
    }
}

